src/main/java/us/kbase/auth/client/internal/TokenCache.java

```diff
@@ -0,0 +1,150 @@
+package us.kbase.auth.client.internal;
+
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import us.kbase.auth.AuthToken;
+
+/**
+ * Caches tokens to avoid network queries to the token provider.
+ * 
+ * Tokens are cached for 5 minutes.
+ * 
+ * Tokens are stored until the size of the cache is greater than the maximum
+ * allowed size. Tokens are then ordered by most recent access and the oldest
+ * tokens are discarded to return the cache to its nominal size.
+ * 
+ * This class is thread safe.
+ * 
+ * @author gaprice@lbl.gov
+ *
+ */
+public class TokenCache {
+	
+	// TODO CODE replace this with Caffeine.
+	
+	/* should really handle tokens as char[] instead of String so the
+	 * char[] can be wiped when it's no longer needed. But since they're going
+	 * over the wire they're probably already strings or will be converted to
+	 * Strings at some point or another.
+	 */
+	
+	// only non-final so can be tested via reflection.
+	private static int MAX_AGE_MS = 5 * 60 * 1000; // 5 min
+	
+	final private int size;
+	final private int maxsize;
+	final private Map<String, UserDate> cache;
+	
+	/**
+	 * Create a new TokenCache.
+	 * @param size the nominal size of the cache in tokens
+	 * @param maxsize the maximum size of the cache in tokens
+	 */
+	public TokenCache(int size, int maxsize) {
```

Comment:
Do we need @param size here? this.size is not used.

Reply:
It's used in putValidToken

Reply:
which line are you referring to?

Reply:
Oh, I saw it.

Reply:
https://github.com/kbase/auth2_client_java/pull/4/files#diff-18a649dcf7643e395e296ddfe7fcd92393b1d5a79d2afa4da7cf88ee036a0275R118

Reply:
👍